### ISSUE TITLE
[Arm64] Crypto assertion failure

### ISSUE THREAD

**Issue by sdmaclea** [dotnet/coreclr#18104]
Recent regression 
JIT/HardwareIntrinsics/Arm64/Crypto/Crypto.sh
```
corerun Crypto.exe
System.Runtime.Intrinsics.Arm.Arm64.Aes.IsSupported = True
System.Runtime.Intrinsics.Arm.Arm64.Sha1.IsSupported = True
System.Runtime.Intrinsics.Arm.Arm64.Sha2.IsSupported = True
Running tests
testCryptoOp<Byte, Vector128`1 >Aes: Check Passed
testCryptoOp<Byte, Vector128`1 >Aes: Check Passed
testCryptoOp<Byte, Vector128`1 >Aes: Check Passed
testCryptoOp<Byte, Vector128`1 >Aes: Check Passed

Assert failure(PID 12709 [0x000031a5], Thread: 12709 [0x31a5]): Assertion failed 'compiler->compFloatingPointUsed || ((mask &amp; RBM_FLT_CALLEE_SAVED) == 0)' in 'Arm64intrisicsTest.Program:TestSha1()' (IL size 437)

    File: src/jit/lsrabuild.cpp Line: 572
```

Failing on 6185b9af8
Passing on fa3e66ea42

@dotnet/arm64-contrib @CarolEidt

**Comment by CarolEidt** [comment]
@sdmaclea - that's probably a safe bet. Sorry about that - I'll take a look.

**Comment by CarolEidt** [comment]
@sdmaclea - this will be a bit tricky for me to track down, as the altjit doesn't recognize the intrinsics, so I'll have to get access to hardware. In the meantime, could you possibly post a jitdump?

**Comment by sdmaclea** [comment]
Sure.  I'll do it in the morning.
There was also an idea to enable setting the CPU feature flags for the altjit/jit through the environment.

**Comment by sdmaclea** [comment]
Looks like this is passing on b23f821.  This doesn't make sense if it was failing on 6185b9a.  I'll try to reproduce...

**Comment by sdmaclea** [comment]
@CarolEidt Here is the JitDump in the failing case
[jitDump.zip](https://github.com/dotnet/coreclr/files/2036239/jitDump.zip)

**Comment by CarolEidt** [comment]
@sdmaclea thanks! It doesn't seem to be failing for me - I'll have a look at the jitdump, and then if that doesn't elucidate, I'll check with exactly the build that you say is failing, and make sure I've got the latest test build (I'm currently using an old build of the tests).

**Comment by CarolEidt** [comment]
@sdmaclea - I've figured out why it's not failing for me - the intrinsics are being treated, even on the arm64 system I'm using - as not supported. While I'm trying to track that down, I have an idea about what the issue might be, though I don't yet understand why it would fail now, and not before.

On xarch, the `compFloatingPointUsed` flag is set here: https://github.com/dotnet/coreclr/blob/master/src/jit/hwintrinsicxarch.cpp#L794 if the `HW_Flag_NoFloatingPointUsed` flag is *not* set for the intrinsics (I hate those double negatives!). This flag must be set prior to register allocation if floating point registers are used - otherwise it doesn't initialize its data structures properly to handle them.

It appears that this is not being done on ARM64, and it *may* be the case that the generated code has changed slightly, such that before there was another use of floating point in the method, which is no longer there.

**Comment by sdmaclea** [comment]
@tannergooding mentioned that to me in February.  I though I added someplace like `impSIMDPopStack()`, but I don't see it on the tip.

**Comment by sdmaclea** [comment]
@CarolEidt Crypto extensions are not available on all arm64 platforms.  They are optional ARMv8 features.

**Comment by sdmaclea** [comment]
#18118 did not fix the failure.....

**Comment by CarolEidt** [comment]
Right - `compFloatingPointUsed` needs to be set for any intrinsic that will read or write any floating point registers, which #18118 didn't do.

**Comment by sdmaclea** [comment]
> Which #18118 didn't do.

Are you aware of a specific case?  It should have set it for any intrinsic which read or wrote SIMD.  I don't remember any intrinsic which read or wrote floating point which wasn't simd.

**Comment by CarolEidt** [comment]
If you look here: https://github.com/dotnet/coreclr/blob/master/src/jit/lsraarm64.cpp#L1080, this instruction is using a float internal register, but it does not produce a SIMD result.

**Comment by sdmaclea** [comment]
I just looked through it.  That does look like the only current exception.  To #18118.  Shall I amend #18118.

**Comment by sdmaclea** [comment]
I guess I should because I can test it.

**Comment by sdmaclea** [comment]
Now #18118 fixes the issue.  

`compFloatingPointUsed` & `_usesSIMDTypes` seem to overlap.  I wonder if `compFloatingPointUsed` should be set in setUsesSIMDTypes(true) case.

**Comment by CarolEidt** [comment]
> compFloatingPointUsed & _usesSIMDTypes seem to overlap. I wonder if compFloatingPointUsed should be set in setUsesSIMDTypes(true) case.

I think that's probably true, though it's not sufficient because there are intrinsics (such as the one at issue here) that use floating point but don't use SIMD types. So I'm not sure how much that buys.

**Comment by CarolEidt** [comment]
Fixed with #18118
